### ISSUE TITLE
Error reporting from calculate_observables

### ISSUE THREAD

**Issue by pathron** [FlexibleSUSY/FlexibleSUSY#195]
Douglas found that when an exception was thrown from code called by calculate_observables it is not passed on to the slha output file when using the run_<modelname>.cpp interface. 

This behaviour is clear because in templates/run.cpp.in the problems are extracted in line 80 before the observables are calculated, and is then passed without further modification to SPINFO block in line 100.  However in addition to this problems that are flagged in the models object during by calculate_observables are not visible in run_\<modelname\>.cpp.  This is because run_\<modelname\>.cpp calls the calculate_observables function with 4 arguments, which creates a local copy and runs that to the specified scale before passing the local copy to the three argument calculate_observables.

We need to decide:

1) How we want to report errors in observables
2) How to implement this

For 1) I think we should *not* report errors from the observables in the SPINFO block since the spectrum is not the problem and the user may want to differentiate.  I think we should create something like an OBSINFO block or preferably one for each observable since we will add more observables in future.  Then we can separate the calculation of each observable, only flag observables that the problem is relevant for and allow the user select which observables are calculated.  Anyway either way we should then provide clear instructions to the to check this if they use the observables output.

The specific error itself was appearing because Douglas had a buggy model file, which we just fixed, and now it is gone. So unfortunately I don't have an example of this happening or test fixes, but I think it is clear enough that we should report errors from observables and we can manually trigger them with temporary hacks if need be to test that the output works.

**Comment by Expander** [comment]
Hi Peter and Douglas,

thanks for reporting this issue!

I principle I like the idea to record the problems for each observable separately. 

## C++ interface

At the C++ level we may use our general `Spectrum_generator_problems`, which is intended to collect all problems, e.g. problems from the BVP solver and from each model. It may in addition collect the problems from each observable in form of a `vector<Observable_probles>` or similar. The `calculate_observables` function may fill this vector of problems.

## SLHA output

What do you think would be a convenient structure for reporting the observable problems to the SLHA file? Out of my head I could think of various ways, for example creating one `OBSINFO` for each observable of the form
~~~
Block OBSINFO
   0 aMuon           # observable name
   1 FlexibleSUSY    # spectrum generator name
   2 2.4.1           # spectrum generator version
   3 XYZ imprecise   # warning
   3 UVW imprecise   # warning
   4 non-perturbative running # error
Block OBSINFO
   0 h->gg           # observable name
   1 FlexibleSUSY    # spectrum generator name
   2 2.4.1           # spectrum generator version
   3 XYZ imprecise   # warning
   3 UVW imprecise   # warning
   4 non-perturbative running # error
~~~
Another possibility would be for example:
~~~
Block OBSINFO aMuon  # name of observable
   3 XYZ imprecise   # warning
   3 UVW imprecise   # warning
   4 non-perturbative running # error
Block OBSINFO h->gg  # name of observable
   3 XYZ imprecise   # warning
   3 UVW imprecise   # warning
   4 non-perturbative running # error
~~~
Another possibility would be for example:
~~~
Block OBSINFO
   3 aMuon XYZ imprecise    # warning
   3 aMuon UVW imprecise    # warning
   4 aMuon non-perturbative running # error
   3 h->gg XYZ imprecise    # warning
   3 h->gg UVW imprecise    # warning
   4 h->gg non-perturbative # error
~~~

## Mathematica level

At the Mathematica level we have the function `<Model>GetProblems` which returns a list of problems. We may simply add observable-specific problems to this list:
~~~
SMGetProblems[]
~~~
would output:
~~~
{
  TwoScaleSolver -> { NoConvergence -> True },
  SM -> { Tachyons -> { M[hh], M[Ah] } },
  aMuon -> { NonPerturbative -> True }
}
~~~

**Comment by pathron** [comment]
Thanks for the very detailed response!

### **At the SLHA level:**

* l think the first possibility creates an ambiguity in OBSINFO[1] etc.  I guess the idea would be the user checks entry 0 as well as the block name to determine what the other entries mean.  However I think that is not within the usual spirit of slha and maybe too complicated.  I see that your thinking is to preserve entries 3 and 4 as the warnings and errors.  

* I also don't like the third possibility you show.  the user then has to check what the string says before they know which calculation has been invalidated

* I think I like the second example better, however I think having a block name with two separate parts may create confusion, e.g. in python people may do `blockname = line.split()[1]` to get the blockname and I don't see the advantage of making them `check line.split()[1]` and `line.split()[2]` to determine the meaning of the entries since just checking `line.split()[1]` will never be useful, so I suggest just making the observable part of the block name, either OSBINFO-aMuon or even just AMUINFO or similar.

* Another possibility if you are worried about a proliferation of INFO blocks is to have double indices here, e.g.

```
Block OBSINFO
   1   3  aMuon XYZ imprecise    # warning
   1   3  aMuon UVW imprecise    # warning
   1   4  aMuon non-perturbative running # error
   2   3  h->gg XYZ imprecise    # warning
   2   3  h->gg UVW imprecise    # warning
   2   4  h->gg non-perturbative # error
```

### **At the C++ level:**

This sounds reasonable.  However `calculate_observables` currently has just the `CMSSM_mass_eigenstates` as an argument for getting results from the spectrum generator.  So we would need to change this.  Also technically one may argue that calculating observables is not part of the spectrum generation so having the observables errors stored there may be a bit strange.     Alternatively maybe we could use the `CMSSM_observables` struct and possibly promote this to a class?   `calculate_observables` could also then be a method of this class instead of a function though I am not sure what the design benefits / disadvantages there are to that.

I am fine with with your suggestion for the mathematica level approach.

**Comment by Expander** [comment]
Hi Peter,

many thanks for your feedback!

## SLHA level

I absolutely agree with your opinion, my suggestions 1 and 3 are not good. I like your two suggestions:

a) using a unique name for each observable, like `OBSINFO-aMuon` and `OBSINFO-h->gg`
b) using a unique index for each observable in a `OBSINFO` block:
~~~
Block OBSINFO
   1   3  aMuon XYZ imprecise    # warning
   1   3  aMuon UVW imprecise    # warning
   1   4  aMuon non-perturbative running # error
   2   3  h->gg XYZ imprecise    # warning
   2   3  h->gg UVW imprecise    # warning
   2   4  h->gg non-perturbative # error
~~~
Personally I like b), because we have fewer blocks which looks cleaner to me. Also concerning filtering the SLHA output with scripts the approach b) looks easier to me. 

In any case the user must know either of
a) the name of the observable (`aMuon` or `h->gg`)
b) the unique number of the observable (1 or 2)

## C++ level

I like your approach of adding a dedicated error class to the `@ModelName@_observables` struct. In this way the results for the observables and the errors are closely related and both are returned from the function.

After the observables struct has been returned, the question is how to proceed with the errors. Initially I thought one would add the observable error class to the spectrum generator error class, so all errors are at one place and can be written to the output:
~~~C++
auto problems = spectrum_generator.get_problems();
auto obs = calculate_observables(...);
problems.add(obs.problems);
// write all problems to the output
slha_io.set_spinfo(problems);
~~~
But this is a bit misleading, because `set_spinfo()` should only modify the `SPINFO` block.

An alternative and backwards-compatible approach would be:
~~~
auto problems = spectrum_generator.get_problems();
slha_io.set_spinfo(problems);

auto obs = calculate_observables(...);
slha_io.set_extra(std::get<0>(models), scales, obs);
~~~
The function `set_extra()` already exists and would be called in exactly the same way.
However, `obs` would now also contain the corresponding errors. I think this is a nicer approach than the first one.

What is your opinion?

**Comment by pathron** [comment]
Yes I like the idea of using set_extra() to set the errors as well, it is simple and also clear. Using set_spinfo() would be confusing I think.  If I understand your suggestion correctly `calculate_observables` remains a function, but now the observables return value will also contain the problems as that will be a part of the `@ModelName@_observables` class?  Looks good to me. Should there be just one problems for the observables with different flags naming the observable and the problem?  or a separate problems class for each observable? 

Regarding the slha file I am fine if you prefer option b).

**Comment by Expander** [comment]
> If I understand your suggestion correctly `calculate_observables` remains a function, but now the observables return value will also contain the problems as that will be a part of the `@ModelName@_observables` class?

Yes, this was my proposal. :)

> Should there be just one problems for the observables with different flags naming the observable and the problem? or a separate problems class for each observable?

I think it would be good to have one class/struct for each observable, which are collected in an overall "observable problems" class. Each observable may have its own specific error types, so I think this would be more flexible. 

Would it be ok if I start an implementation of this and create a PR for review?

**Comment by Expander** [comment]
A short question: Which kind of error did you and Douglas see? Was it a non-perturbative running or a non-perturbative numerical value for an observable?

**Comment by pathron** [comment]
It was a non-perturbative running error.

**Comment by Expander** [comment]
Thanks for the information!
I've added PR #214 where I will perform the development. I'll come back to you when I'm ready. :)

**Comment by Expander** [comment]
Hi Peter,

I have a first working version, which implements the scheme we have discussed. Would you mind having a look? Below you'll find an SLHA input file which triggers the output of an `OBSINFO` block.

I've written the extension with the following in mind:

* PR #214 should not increase the compile time (no new headers should be included). This led to the use of output iterators, instead of returning the problems in form of a `std::vector`. (We could avoid using `std::vector` and output iterators are more general.)
* Modifying/adding new observable problems should not require large modification of the templates and the meta code, in order to minimize annoyance by re-generation and re-compilation of many files.

Currently there are only two types of observable problems:
* general problems, for example non-perturbative running to the scale where the observables are to be calculated
* `a_muon` problems, for example non-perturbative running in the `run_to_MSUSY()` routine.

P.S.: One benefit of this modification is that we can now make the `@ModelName@_mass_eigenstates&`, which is passed to `calculate_observables()`, a `const` reference. This nicely indicates to the caller that the model will not be modified.

P.P.S.: Looking at `templates/librarylink.cpp.in` I see further possiblities for refactorings to make the code easier to extend and faster to compile regarding the other problem classes. Probably this should be done in a separate PR.

**Comment by Expander** [comment]
~~~
Block MODSEL                 # Select model
#   12    1000                # parameter output scale (GeV)
Block FlexibleSUSY
    0   1.000000000e-04      # precision goal
    1   0                    # max. iterations (0 = automatic)
    2   0                    # algorithm (0 = all, 1 = two_scale, 2 = semi_analytic)
    3   1                    # calculate SM pole masses
    4   4                    # pole mass loop order
    5   4                    # EWSB loop order
    6   4                    # beta-functions loop order
    7   4                    # threshold corrections loop order
    8   1                    # Higgs 2-loop corrections O(alpha_t alpha_s)
    9   1                    # Higgs 2-loop corrections O(alpha_b alpha_s)
   10   1                    # Higgs 2-loop corrections O((alpha_t + alpha_b)^2)
   11   1                    # Higgs 2-loop corrections O(alpha_tau^2)
   12   1                    # force output
   13   3                    # Top pole mass QCD corrections (0 = 1L, 1 = 2L, 2 = 3L)
   14   1.000000000e-11      # beta-function zero threshold
   15   1                    # calculate observables (a_muon, ...)
   16   0                    # force positive majorana masses
   17   0                    # pole mass renormalization scale (0 = SUSY scale)
   18   0                    # pole mass renormalization scale in the EFT (0 = min(SUSY scale, Mt))
   19   0                    # EFT matching scale (0 = SUSY scale)
   20   2                    # EFT loop order for upwards matching
   21   1                    # EFT loop order for downwards matching
   22   0                    # EFT index of SM-like Higgs in the BSM model
   23   1                    # calculate BSM pole masses
   24   124111421            # individual threshold correction loop orders
   25   0                    # ren. scheme for Higgs 3L corrections (0 = DR, 1 = MDR)
   26   1                    # Higgs 3-loop corrections O(alpha_t alpha_s^2)
   27   1                    # Higgs 3-loop corrections O(alpha_b alpha_s^2)
   28   1                    # Higgs 3-loop corrections O(alpha_t^2 alpha_s)
   29   1                    # Higgs 3-loop corrections O(alpha_t^3)
   30   1                    # Higgs 4-loop corrections O(alpha_t alpha_s^3)
   31   0                    # loop library (0 = softsusy)
Block FlexibleSUSYInput
    0   0.00729735           # alpha_em(0)
    1   125.09               # Mh pole
Block SMINPUTS               # Standard Model inputs
    1   1.279340000e+02      # alpha^(-1) SM MSbar(MZ)
    2   1.166378700e-05      # G_Fermi
    3   1.184000000e-01      # alpha_s(MZ) SM MSbar
    4   9.118760000e+01      # MZ(pole)
    5   4.200000000e+00      # mb(mb) SM MSbar
    6   1.733400000e+02      # mtop(pole)
    7   1.777000000e+00      # mtau(pole)
    8   0.000000000e+00      # mnu3(pole)
   11   5.109989020e-04      # melectron(pole)
   12   0.000000000e+00      # mnu1(pole)
   13   1.056583570e-01      # mmuon(pole)
   14   0.000000000e+00      # mnu2(pole)
   21   4.750000000e-03      # md(2 GeV) MS-bar
   22   2.400000000e-03      # mu(2 GeV) MS-bar
   23   1.040000000e-01      # ms(2 GeV) MS-bar
   24   1.270000000e+00      # mc(mc) MS-bar
Block MINPAR                 # Input parameters
    1   0.192                # Lambda(Qin)
Block EXTPAR                 # Input parameters
    0   1000                 # input scale Qin
    1   0                    # scale QEWSB
~~~

**Comment by Expander** [comment]
@pathron would you mind trying to run the SM with this SLHA file to see if the output is OK?
~~~
models/SM/run_SM.x --slha-input-file=slha.in
~~~

**Comment by pathron** [comment]
OK I get 
```
Block OBSINFO
     0     3   general: non-perturbative running to scale 0.000000 GeV
Block FlexibleSUSYLowEnergy Q= 1.00000000E+03
    21     0.00000000E+00   # Delta(g-2)_muon/2 FlexibleSUSY
Block EFFHIGGSCOUPLINGS
       25       22       22     0.00000000E+00   # Abs(effective H-Photon-Photon coupling)
       25       21       21     0.00000000E+00   # Abs(effective H-Gluon-Gluon coupling)
```
I wonder though looking at this, is the correspondence between OBSINFO[0,3] and FlexibleSUSYLowEnergy[21] intuitive? If i renad this I would not be sure which observable or observables this applies to, especially since in this case its not really an observable specific type of error.

**Comment by pathron** [comment]
would it make sense and be general enough to match the number of the OBSINFO  entry with the first number in the corresponding  entry in FlexibleSUSYLowEnergy and then if it apples to many observables you have report it for each.  I am bit confused by what happens with the effective couplings though since they are reported in FlexibleSUSYLowEnergy but in their own block.

**Comment by Expander** [comment]
Thanks for your comments!

The error that you see in the `OBSINFO` block does not come from the `aMuon` calculation. It comes from the failing RG running, which we do before we start calculating the observables. This is why it says `general: ...`. I assigned the special index 0 to thes "general" errors.

A problem in the `aMuon` calculation would read:
~~~
Block OBSINFO
     1     3   aMuon: non-perturbative running to scale 0.000000 GeV
~~~
However, for this SLHA input file a "general" problem appears earlier, so the general problem is reported only (as the program stops after this general problem).

Currently, I've assigned the (arbitrary) index `1` to `aMuon`. I was hoping that the description in the `OBSINFO` block makes clear which observable has a problem.

Your idea of using the same index for the error as for the observable itself is good, but may lead to dis-ambiguities: For example, if a user chooses the same index for two different observables in different blocks:
~~~
   {FlexibleSUSYLowEnergy,
           {{21, FlexibleSUSYObservable`aMuon}}
   } ,
   {GM2CalcLowEnergy,
           {{21, FlexibleSUSYObservable`aMuonGM2Calc}}
   },
~~~
In this case the corresponding output in `OBSINFO` would be:
Block OBSINFO
     21    3   aMuon: non-perturbative running to scale 0.000000 GeV
     21    3   aMuonGM2Calc: non-perturbative running to scale 0.000000 GeV
~~~
with the same index.

**Comment by pathron** [comment]
yes that is why i said i was confused by the effectivecouplings case, if all observables were repoorted in the same block it would work, but i guess that is not possible because they could also have different structures  like EFFHIGGSCOUPLINGS.  yes. 

I guess we may have to rely on the comments making it clear to the user what the entries mean unless they check the documentation.  but in that snese I am not sure general: is clear?  What about "all observables: " assuming this error will indeed invalidate all of them?

**Comment by Expander** [comment]
Good point, "general" is probably not clear. I will change it to "all observables", which is much clearer. :)

**Comment by Expander** [comment]
I've changed it in commit bc35d772e16b9ac . Is there something else that could be improved?

**Comment by pathron** [comment]
OK sorry for not replying earlier  I now get this

```
Block OBSINFO
     0     3   all observables: non-perturbative running to scale 0 GeV
Block FlexibleSUSYLowEnergy Q= 1.00000000E+03
    21     0.00000000E+00   # Delta(g-2)_muon/2 FlexibleSUSY
Block EFFHIGGSCOUPLINGS
       25       22       22     0.00000000E+00   # Abs(effective H-Photon-Photon coupling)
       25       21       21     0.00000000E+00   # Abs(effective H-Gluon-Gluon coupling)

```

I think its good to go.  Thanks!
